### PR TITLE
enhance: adding an optimized flag

### LINKED ISSUE #25015: Add `optimized` flag to Keycloak CR to indicate if custom image should be treated as optimized
### Description

Currently, whenever a custom Keycloak image is provided, the Operator expects it to be [optimized](https://www.keycloak.org/server/containers#_creating_a_customized_and_optimized_container_image) as it is considered a best practice. However, this does not align with some use cases when it's desirable to have an unoptimized custom image, e.g. when some build time options are unknown at the image build time, and need to be specified in the Keycloak CR.

We should add an optional switch to Keycloak CR to indicate if a custom image was optimized. This option would work  only with custom images, not with the default image.

### Discussion

_No response_

### Motivation

_No response_

### Details

_No response_

### PR DESCRIPTION
**Pull request by shawkins**
@vmuzikar do you want a release note about this?

closes: #25015

### CODE REVIEW

**Review by vmuzikar — COMMENTED** [review]
> do you want a release note about this?

Yeah, I think it's worth noting.

And I think we need to update [docs](https://github.com/keycloak/keycloak/blob/0b48bef0b102108a0bd002f99562766553030013/docs/guides/operator/customizing-keycloak.adoc) as well.

File: operator/src/main/java/org/keycloak/operator/crds/v2alpha1/deployment/KeycloakSpec.java
Comment (shawkins) [reply operator/src/main/java/org/keycloak/operator/crds/v2alpha1/deployment/KeycloakSpec.java:48]: I'm not sure that is much clearer as you still have to be aware of the image property to know the full meaning - because you would rightfully know there's always an image reference on the StatefulSet.

You'd have to do something like customImageOptimized for the property name alone to be sufficiently descriptive - and then it's much too verbose / and uses a different convention that the image property.  I'd be fine with leaving it as just optimized.  

It's not handled here, but there's is also a chance that some users may want to change the base image referenced by the operator to be an optimized one, in which case this property could override the default non-optimized behavior to optimized as well.

Comment (vmuzikar) [reply operator/src/main/java/org/keycloak/operator/crds/v2alpha1/deployment/KeycloakSpec.java:48]: My reasoning was to align with other similarly named fields in pod templates, e.g. `image` and `imageOptimized`. Only `optimized` might make the impression that it's something related to KC instance rather than the image itself.

> It's not handled here, but there's is also a chance that some users may want to change the base image referenced by the operator to be an optimized one, in which case this property could override the default non-optimized behavior to optimized as well.

Yeah, the `optimized` flag might work for the default image as well. If not explicitly specified, we'd default to the current behaviour (`false` for default image, `true` for custom). We'd document this behaviour.

File: operator/src/test/java/org/keycloak/operator/testsuite/unit/PodTemplateTest.java
Comment (vmuzikar) [reply operator/src/test/java/org/keycloak/operator/testsuite/unit/PodTemplateTest.java:1]: I think we're not testing the case when a custom image is specified.

File: operator/src/main/java/org/keycloak/operator/crds/v2alpha1/deployment/KeycloakSpec.java
Comment (shawkins) [reply operator/src/main/java/org/keycloak/operator/crds/v2alpha1/deployment/KeycloakSpec.java:48]: To support this, there's an additional config value as well.

File: docs/documentation/release_notes/topics/24_0_0.adoc
Comment (vmuzikar) [inline docs/documentation/release_notes/topics/24_0_0.adoc:29]: I'd try to sell it a bit more.
```suggestion
The Keycloak CR now includes an optimized field, which may be used to override the default assumption about whether the image is optimized. As a result, you can use the CR to configure build time options also when a custom Keycloak image is used.
```

File: docs/guides/operator/customizing-keycloak.adoc
Comment (vmuzikar) [inline docs/guides/operator/customizing-keycloak.adoc:53]:
```suggestion
While it is considered a best practice to optimize custom image, if you want to use a non-optimized custom image and incur the overhead of re-augmentation - that is still possible. This is useful when you want to configure some build time options via the Keycloak CR. You just need set the `optimzed` field to `false` as shown in this example:
```

**Review by vmuzikar — COMMENTED** [review]
@shawkins Thanks for the updates.

Code changes LGTM. Added two minor comments.

One thing I'm still not sure about is the name of the option (as I stated above). While `optimized` is aligned with the actual CLI option name, I think in the context of the CR it's more related to the image, and not to the Keycloak instance/configuration itself which calling it just `optimized` implies to me. In other words, you cannot use the `optimized` flag on its own, the image needs to be optimized and that needs to be reflected by setting that flag. So it's basically an image related flag. Hence I'd call it `imageOptimized`. But I don't feel strongly about it, so not blocking for me, and I'll let @mabartos and @Pepo48 to chime in.

**Comment by shawkins** [comment]
I'll resolve the DCO after we agree on the naming convention - seems like github should know to sign suggested edits :(

**Review by vmuzikar — DISMISSED** [review]
LGTM, thanks @shawkins.

Before merging this, I'd like to give the opportunity for @mabartos and @Pepo48 to chime in as well, especially around the option name.

**Review by mabartos — COMMENTED** [review]
@shawkins I'm more inclined towards @vmuzikar's opinion related to the name of the field. It makes more sense to me in the context of the CR configuration as it's related to the Keycloak image itself and probably emphasize more the purpose of the optimized image.

I could imagine two different flows of thoughts:
1. ('imageOptimized') - asking myself "Hmm, did I specify an optimized image?"
2. ('optimized') - gives me feeling like: "Do you want to have the instance optimized?" 

Of course, I need to read docs about it, and those might be quite silly assumptions, but it's just my initial feeling about the wording xP
 
Even in the Pod and SS object, these fields will be positioned next to each other(`image`, `imagePullPolicy`, `imageOptimized`) - tiny benefit.

But both approaches work for me. If you think it's better to have it `optimized`, leave it as is :)

File: docs/documentation/release_notes/topics/24_0_0.adoc
Comment (shawkins) [reply docs/documentation/release_notes/topics/24_0_0.adoc:29]: My qualm with imageOptimized is that the field applies whether or not an explicit image is set.

How about a third option - startOptimized - that may feel less generic.

> Even in the Pod and SS object, these fields will be positioned next to each other(image, imagePullPolicy, imageOptimized) - tiny benefit.

optimized will actually end up in the command args.

File: docs/guides/operator/customizing-keycloak.adoc
Comment (mabartos) [inline docs/guides/operator/customizing-keycloak.adoc:53]:
```suggestion
While it is considered a best practice to optimize custom image, if you want to use a non-optimized custom image and incur the overhead of re-augmentation - that is still possible. This is useful when you want to configure some build time options via the Keycloak CR. You just need set the `startOptimized` field to `false` as shown in this example:
```

File: docs/guides/operator/customizing-keycloak.adoc
Comment (mabartos) [inline docs/guides/operator/customizing-keycloak.adoc:64]:
```suggestion
  startOptimized: false
```

**Review by mabartos — COMMENTED** [review]
> One last consideration on this is that some users may find it odd that they have to control this flag manually for the primary usecase - if you supply any build time option we could automatically interpret that as a non-optimized scenario.

@shawkins IMHO, I would not complicate things. The standard way is to assuming the image is already augmented. The ability to use non-augmented image is kinda above standard and who wants to use it, just specify the flag :P WDYT?

**Comment by shawkins** [comment]
> @shawkins IMHO, I would not complicate things. The standard way is to assuming the image is already augmented. The ability to use non-augmented image is kinda above standard and who wants to use it, just specify the flag :P WDYT?

Based upon the issues we've seen it does appear that usage via the operator abstracts just enough to make things even more confusing wrt --optimized.  While it's understandable for us to push the responsibility to users, it's not great in terms of usability.  Specifying build-time options via non-cli options with --optimized, is a soft spot for the server.  We will now warn in the log about, but that still means you have to check the log.

File: operator/src/main/java/org/keycloak/operator/crds/v2alpha1/deployment/KeycloakSpec.java
Comment (vmuzikar) [inline operator/src/main/java/org/keycloak/operator/crds/v2alpha1/deployment/KeycloakSpec.java:47]:
```suggestion
    @JsonPropertyDescription("Set to force the behavior of the --optimized flag for the start command. If left unspecified the operator will assume custom images have already been augmented.")
```

**Review by vmuzikar — APPROVED** [review]
LGTM, thanks @shawkins.

I pushed one small change to remove an extra space in the option description.

**Comment by shawkins** [comment]
> LGTM, thanks @shawkins.
> 
> I pushed one small change to remove an extra space in the option description.

That extra space is a hard habit to break - was required to do that all the back in high school and can't seem to quit.

**Comment by Pepo48** [comment]
LGTM. Thanks, @shawkins.

**Review by mabartos — APPROVED** [review]
@shawkins LGTM in this form for now. 

It'd be good to discuss the topic related to the offline discussion about forcing the mode if `startOptimized` is explicitly specified vs. automatically determining the mode based on the provided configuration, as the docs would change in that case. 

I'd keep it as is for now as it's a nice improvement; we can make another one later. 

@shawkins Please squash commits and sign them.

**Comment by shawkins** [comment]
> It'd be good to discuss the topic related to the offline discussion about forcing the mode if startOptimized is explicitly specified vs. automatically determining the mode based on the provided configuration, as the docs would change in that case.

That was captured as #25391

> @shawkins Please squash commits and sign them.

Looking forward to when the suggested edits are more automated again...

**Pull request merged**
